### PR TITLE
Fixed out-of-bounds read in parallel version of ippGaussianBlur()

### PR DESCRIPTION
**Pull request by eplankin**
Fixed issue [#11303](https://github.com/opencv/opencv/issues/11303).
The problem was the following: in the parallel version of the function in case when kernelSize/2 is bigger than a height of a tile,  out-of-bounds read happens when the second and the one before the last tiles are processed.
Now sequential mode is used in such cases.

### CODE REVIEW

**Comment by alalek** [comment]
As a "bugfix" this patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into 4.x regularly (weekly/bi-weekly).

Please:
- change "base" branch of this PR: 4.x => 3.4 (use "Edit" button near PR title)
- rebase your commits from 4.x onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/4.x`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

**Comment by alalek** [comment]
Problem is confirmed (4+ threads):

```
$ valgrind ./bin/opencv_test_imgproc --gtest_filter=*regression_11303 --test_threads=4
...
[ RUN      ] Imgproc_GaussianBlur.regression_11303
==697273== Thread 3:
==697273== Invalid read of size 32
==697273==    at 0x592CA8D: icv_l9_ownFilterGaussianRow_Brd_32f_C1 (in /home/alalek/projects/opencv/build/opencv/lib/libopencv_imgproc.so.3.4.17)
==697273==    by 0x59392FD: icv_l9_ippiFilterGaussian_32f_C1R_L (in /home/alalek/projects/opencv/build/opencv/lib/libopencv_imgproc.so.3.4.17)
==697273==    by 0x504436F: llwiFilterGaussian_Process (iw_image_filter_gaussian.c:385)
==697273==    by 0x50441AF: llwiFilterGaussian_ProcessWrap (iw_image_filter_gaussian.c:350)
==697273==    by 0x50437CF: llwiFilterGaussian_ProcessWrap (iw_image_filter_gaussian.c:241)
==697273==    by 0x50433F0: iwiFilterGaussian (iw_image_filter_gaussian.c:113)
==697273==    by 0x4D42A87: ipp::iwiFilterGaussian(ipp::IwiImage const&, ipp::IwiImage&, int, double, ipp::IwiFilterGaussianParams const&, ipp::IwiBorderType const&, ipp::IwiTile const&) (iw_image_filter.hpp:281)
==697273==    by 0x4D58E60: cv::ipp_gaussianBlurParallel::operator()(cv::Range const&) const (smooth.dispatch.cpp:507)
==697273==    by 0x72042D4: cv::(anonymous namespace)::ParallelLoopBodyWrapper::operator()(cv::Range const&) const (parallel.cpp:340)
==697273==    by 0x7206310: cv::ParallelJob::execute(bool) (parallel_impl.cpp:344)
==697273==    by 0x7206711: cv::WorkerThread::thread_body() (parallel_impl.cpp:468)
==697273==    by 0x7205F5D: cv::WorkerThread::thread_loop_wrapper(void*) (parallel_impl.cpp:284)
==697273==  Address 0x1265d864 is 1,785,108 bytes inside a block of size 1,785,132 alloc'd
==697273==    at 0x484486F: malloc (vg_replace_malloc.c:381)
==697273==    by 0x6F85D23: cv::fastMalloc(unsigned long) (alloc.cpp:150)
==697273==    by 0x712EFDE: cv::StdMatAllocator::allocate(int, int const*, int, void*, unsigned long*, int, cv::UMatUsageFlags) const (matrix.cpp:147)
==697273==    by 0x71317B3: cv::Mat::create(int, int const*, int) (matrix.cpp:645)
==697273==    by 0x71310B5: cv::Mat::create(int, int, int) (matrix.cpp:537)
==697273==    by 0x71300FB: cv::Mat::Mat(cv::Size_<int>, int, cv::Scalar_<double> const&) (matrix.cpp:371)
==697273==    by 0x64EC88: opencv_test::(anonymous namespace)::Imgproc_GaussianBlur_regression_11303_Test::Body() (test_filter.cpp:2364)
==697273==    by 0x64EB59: opencv_test::(anonymous namespace)::Imgproc_GaussianBlur_regression_11303_Test::TestBody() (test_filter.cpp:2358)
==697273==    by 0x76DBF0: void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (ts_gtest.cpp:3917)
==697273==    by 0x767FCB: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (ts_gtest.cpp:3953)
==697273==    by 0x74C857: testing::Test::Run() (ts_gtest.cpp:3991)
==697273==    by 0x74D188: testing::TestInfo::Run() (ts_gtest.cpp:4167)
```

File: modules/imgproc/CMakeLists.txt [hunk 1]
```diff
@@ -14,7 +14,7 @@ ocv_add_dispatched_file(undistort SSE2 AVX2)
 ocv_define_module(imgproc opencv_core WRAP java python js)
 
 ocv_check_environment_variables(OPENCV_IPP_GAUSSIAN_BLUR)
-option(OPENCV_IPP_GAUSSIAN_BLUR "Enable IPP optimizations for GaussianBlur (+8Mb in binary size)" OFF)
+option(OPENCV_IPP_GAUSSIAN_BLUR "Enable IPP optimizations for GaussianBlur (+8Mb in binary size)" ON)
```
Comment (alalek) [inline modules/imgproc/CMakeLists.txt:17]: Please revert this change before merge

File: modules/imgproc/test/test_filter.cpp [hunk 2]
```diff
@@ -2355,5 +2355,16 @@ TEST(Imgproc, filter_empty_src_16857)
...
 
+TEST(Imgproc_GaussianBlur, regression_11303)
+{
+    cv::Mat dst;
+    int width = 2115;
+    int height = 211;
+    double sigma = 8.64421;
+    cv::Mat src(cv::Size(width, height), CV_32F, 1);
+    cv::GaussianBlur(src, dst, cv::Size(), sigma, sigma);
+    EXPECT_EQ(0, cv::norm(src, dst, NORM_L2));
```
Comment (alalek) [inline modules/imgproc/test/test_filter.cpp:2366]: This check may fail.

I observed problem by this run:
```
valgrind ./bin/opencv_test_imgproc --gtest_filter=*regression_11303 --test_threads=3
...
Intel(R) IPP version: ippIP AVX2 (l9) 2020.0.0 Gold (-) Oct 19 2019
...
[ RUN      ] Imgproc_GaussianBlur.regression_11303
/home/alalek/projects/opencv/dev/modules/imgproc/test/test_filter.cpp:2366: Failure
Expected equality of these values:
  0
  cv::norm(src, dst, NORM_L2)
    Which is: 7.96355e-05
[  FAILED  ] Imgproc_GaussianBlur.regression_11303 (5123 ms)
```

Also check fails in Debug builder (without IPP):

```
[ RUN      ] Imgproc_GaussianBlur.regression_11303
/build/precommit_linux64_no_opt/3.4/opencv/modules/imgproc/test/test_filter.cpp:2366: Failure
Expected equality of these values:
  0
  cv::norm(src, dst, NORM_L2)
    Which is: 7.96355e-05
[  FAILED  ] Imgproc_GaussianBlur.regression_11303 (252 ms)
```

This check is very strong. We could relax this, e.g:

```
EXPECT_LE(cv::norm(src, dst, NORM_L2), 1e-3);
```

Comment (eplankin) [reply modules/imgproc/test/test_filter.cpp]: Fixed

**Comment by alalek** [comment]
> I cannot reproduced

Message above is just a confirmation how to reproduce the problem, so we could validate proposed fix.

---

BTW,

> `--test-threads=5`
> (nthreads=4)

Looks like there are 4 logic CPUs only.

---

> out-of-memory

It is called as "out of bounds" / OOB problem.

---

Please fix/eliminate build warning before merge.

**Comment by alalek** [comment]
Please convert PR from Draft if it is ready (use "Ready for review" button).

**Review by alalek — APPROVED** [review]
LGTM! Thank you :+1:
